BloomFilter.md

```diff
@@ -49,87 +44,185 @@ pushdown even in columns with high cardinality and when space is at a premium.
+and `filter_insert` does not return a value, but does modify the SBBF.
+
+The `filter_insert` operation first uses the most significant 32 bits
+of its argument, modulo the number of blocks, to select a block to
+operate on. It then uses the least significant 32 bits of the argument
+to `filter_insert` as an argument to `block_insert` called on that
+block.
+
+The `filter_check` operation uses the same method as `filter_insert`
+to select a block to operate on, then uses the least significant 32
+bits of its argument as an argument to `block_check` called on that
+block, returning the result.
+
+In the pseudocode below, the modulus operator is represented with the C
+language's "`%`" operator. The "`>>`" operator is used to denote the
+conversion of an unsigned 64-bit integer to an unsigned 32-bit integer
+containing only the most significant 32 bits, and C's cast operator
+"`(unsigned int32)`" is used to denote the conversion of an unsigned
+64-bit integer to an unsigned 32-bit integer containing only the lest
+significant 32 bits.
+
+```
+void filter_insert(SBBF filter, unsigned int64 x) {
+  block b = filter.getBlock((x >> 32) % filter.numberOfBlocks());
+  block_insert(b, (unsigned int32)x)
 }
+```
 
 ```
+boolean filter_check(SBBF filter, unsigned int64 x) {
+  block b = filter.getBlock((x >> 32) % filter.numberOfBlocks());
+  return block_check(b, (unsigned int32)x)
+}
+```
+
+The use of blocks is from Putze et al.'s [Cache-, Hash- and
+Space-Efficient Bloom
+filters](http://algo2.iti.kit.edu/documents/cacheefficientbloomfilters-jea.pdf)
 
-#### Hash Function
-The function used to hash values in the initial implementation is
-[xxHash](https://cyan4973.github.io/xxHash/), using the function XXH64 with a
-seed of 0 and [following the specification version
+To use an SBBF for values of arbitrary Parquet types, we apply a hash
+function to that value - at the time of writing,
+[xxHash](https://cyan4973.github.io/xxHash/), using the function XXH64
+with a seed of 0 and [following the specification version
 0.1.1](https://github.com/Cyan4973/xxHash/blob/v0.7.0/doc/xxhash_spec.md).
 
-#### Build a Bloom filter
-The fact that exactly eight bits are checked during each lookup means that these filters
-are most space efficient when used with an expected false positive rate of about
-0.5%. This is achieved when there are about 11.54 bits for every distinct value inserted
-into the filter.
+#### Sizing an SBBF
 
-To calculate the size the filter should be for another false positive rate `p`, use the
-following formula. The output is in bits per distinct element:
+The `check` operation in SBBFs can return `true` for an argument that
+was never inserted into the SBBF. These are called "false
+positives". There is not a simple closed-form calculation of this
+probability, but empirically, an SBBF that uses 6 bits of space (or
+6/256 blocks) per unique item inserted will have a false positive
```

Comment:
Why is this 6? Each value is set in a block and each block requires setting 8 bits, one in each 32-bit word. So this isn't possible in this spec, right?

Reply:
I guess this could be the average across all blocks, not the actual number of bits set. That's what the next calculation seems to imply. Although we are setting 8 bits per key, the average is around 6 bits per value if you insert 43691 values.

Reply:
Exactly. While eight bits are **set** each time an item is inserted, that doesn't mean that the fraction `bits_of_space / number_of_distinct_items_inserted` is eight. I attempted to convey this by adding the modifier "of space".

Reply:
I think it would be good to have a note about these calculations. I had a lot of trouble understanding what was intended by the 6/256. I recommend showing calculations that relate the total number of bits (blocks * 256) to bits per distinct value and distinct values. That's easier to understand.

Reply:
After a face-to-face discussion, Ryan explained a bit more how this was confusing. I've changed the wording as a result of that conversation to note that bits-per-insert is a derived value.